src/schedule/zephyr_preempt_domain.c

```diff
@@ -0,0 +1,500 @@
+			       pipe_task,
+			       core);
+			return ret;
+		}
+
+		data->enabled_irq = true;
+
+		return 0;
+	}
+
+	/* if this point is reached then that means we weren't able to
+	 * enable DMA IRQ either because data was NULL or the IRQ was
+	 * already enabled even though the Zephyr thread wasn't started
+	 */
+
+	tr_err(&ll_tr, "failed to register pipeline task %p on core %d",
+	       pipe_task,
+	       core);
+
+	return -EINVAL;
+}
+
+static void disable_dma_irq(struct zephyr_preempt_domain_data *data)
+{
+	dma_interrupt_legacy(data->channel, DMA_IRQ_MASK);
+	dma_interrupt_legacy(data->channel, DMA_IRQ_CLEAR);
+	interrupt_clear_mask(data->irq, BIT(data->channel->index));
+	interrupt_disable(data->irq, data);
+	interrupt_unregister(data->irq, data);
+}
+
+static int unregister_dma_irq(struct zephyr_preempt_domain *domain,
+			      struct pipeline_task *pipe_task,
+			      int core)
+{
+	struct zephyr_preempt_domain_data *crt_data;
+	struct dma *crt_dma;
+	int i, j;
+	/* unregister the DMA IRQ only for PPL tasks marked as "registrable"
+	 *
+	 * this is done because, in case of mixer topologies there's
+	 * multiple PPLs having the same scheduling component so there's
+	 * no need to go through this function for all of those PPL
+	 * tasks - only the PPL task containing the scheduling component
+	 * will do the unregistering
+	 *
+	 */
+	if (!pipe_task->registrable)
+		return 0;
+
+	for (i = 0; i < domain->num_dma; i++) {
+		crt_dma = domain->dma_array + i;
+
+		for (j = 0; j < crt_dma->plat_data.channels; j++) {
+			crt_data = &domain->data[i][j];
+
+			/* skip if DMA IRQ wasn't enabled */
+			if (!crt_data->enabled_irq)
+				continue;
+
+			/* skip if channel not owned by current core */
+			if (crt_data->channel->core != core)
+				continue;
+
+			/* skip if channel is still active */
+			if (crt_data->channel->status == COMP_STATE_ACTIVE)
+				continue;
+
+			disable_dma_irq(crt_data);
+
+			crt_data->enabled_irq = false;
+
+			tr_info(&ll_tr, "unregister_dma_irq(): unregistered IRQ for task %p",
+				pipe_task);
+
+			return 0;
+		}
+	}
+
+	/* if this point is reached then something went wrong */
+	return -EINVAL;
+}
+
+static int zephyr_preempt_domain_unregister(struct ll_schedule_domain *domain,
+					    struct task *task,
+					    uint32_t num_tasks)
+{
+	struct zephyr_preempt_domain *zephyr_preempt_domain;
+	struct zephyr_preempt_domain_thread *dt;
+	struct pipeline_task *pipe_task;
+	int ret, core;
+
+	zephyr_preempt_domain = ll_sch_get_pdata(domain);
+	pipe_task = pipeline_task_get(task);
+	core = cpu_get_id();
+	dt = zephyr_preempt_domain->domain_thread + core;
+
+	tr_info(&ll_tr, "zephyr_preempt_domain_unregister()");
+
+	ret = unregister_dma_irq(zephyr_preempt_domain, pipe_task, core);
+	if (ret < 0) {
+		tr_err(&ll_tr, "failed to unregister DMA IRQ for pipe task %p on core %d",
+		       pipe_task,
+		       core);
+		return ret;
+	}
+
+	/* TODO: thread abortion logic goes here */
```

Comment:
This is probably needed for merging, or...? The logic around dt->handler seems a bit suspicious. When unregistering a domain, handler is not clear, but yet in register we use this field to decide whether to start a new thread. Does this handle multiple register/unregister cycles in its current form? I don't think we need to have everything polished to merge this, but just looking for a sane minimal-functionality set...

Reply:
I agree this look a bit shady.

When I did the testing it seemed to work fine without it. The tests consisted of a bunch of `aplay`/`arecord` having a 3s sleep in between them in order to make the FW reboot + some `aplay` while doing a lot of pause pushing. Also, the DMA trace was disabled during the tests. Also, I did a few consecutive `aplay` (without the FW boot) and it seemed to work fine as well.

I can't guarantee that this will not cause problems on systems with multiple cores as I haven't been able to test that scenario.

This probably works after FW reboot (which should be the only problematic case) because of the fact that `dt->handler` is made NULL by preempt domain's initialization function (which from my understanding should be called after every FW boot)